// playcanvas/build/playcanvas.prf/src/scene/gsplat-unified/gsplat-node-cull-render-pass.js
import { SEMANTIC_POSITION } from '../../platform/graphics/constants.js';
import { RenderPassShaderQuad } from '../graphics/render-pass-shader-quad.js';
import { ShaderUtils } from '../shader-lib/shader-utils.js';
import glslGsplatNodeCullingPS from '../shader-lib/glsl/chunks/gsplat/frag/gsplatNodeCulling.js';
import wgslGsplatNodeCullingPS from '../shader-lib/wgsl/chunks/gsplat/frag/gsplatNodeCulling.js';

class GSplatNodeCullRenderPass extends RenderPassShaderQuad {
		setup(boundsSphereTexture, boundsTransformIndexTexture, transformsTexture, totalBoundsEntries, frustumPlanes) {
				this._boundsSphereTexture = boundsSphereTexture;
				this._boundsTransformIndexTexture = boundsTransformIndexTexture;
				this._transformsTexture = transformsTexture;
				this._totalBoundsEntries = totalBoundsEntries;
				this._frustumPlanes = frustumPlanes;
		}
		execute() {
				this.boundsSphereTextureId.setValue(this._boundsSphereTexture);
				this.boundsTransformIndexTextureId.setValue(this._boundsTransformIndexTexture);
				this.transformsTextureId.setValue(this._transformsTexture);
				this.boundsTextureWidthId.setValue(this._boundsSphereTexture.width);
				this.transformsTextureWidthId.setValue(this._transformsTexture.width);
				this.totalBoundsEntriesId.setValue(this._totalBoundsEntries);
				this.frustumPlanesId.setValue(this._frustumPlanes);
				super.execute();
		}
		constructor(device){
				super(device), this._totalBoundsEntries = 0;
				this.shader = ShaderUtils.createShader(device, {
						uniqueName: 'GSplatNodeCulling',
						attributes: {
								aPosition: SEMANTIC_POSITION
						},
						vertexChunk: 'quadVS',
						fragmentGLSL: glslGsplatNodeCullingPS,
						fragmentWGSL: wgslGsplatNodeCullingPS,
						fragmentOutputTypes: [
								'uint'
						]
				});
				this.boundsSphereTextureId = device.scope.resolve('boundsSphereTexture');
				this.boundsTransformIndexTextureId = device.scope.resolve('boundsTransformIndexTexture');
				this.transformsTextureId = device.scope.resolve('transformsTexture');
				this.boundsTextureWidthId = device.scope.resolve('boundsTextureWidth');
				this.transformsTextureWidthId = device.scope.resolve('transformsTextureWidth');
				this.totalBoundsEntriesId = device.scope.resolve('totalBoundsEntries');
				this.frustumPlanesId = device.scope.resolve('frustumPlanes[0]');
		}
}

export { GSplatNodeCullRenderPass };
