// playcanvas/build/playcanvas.prf/src/scene/gsplat-unified/gsplat-interval-compaction.js
import { Compute } from '../../platform/graphics/compute.js';
import { Shader } from '../../platform/graphics/shader.js';
import { StorageBuffer } from '../../platform/graphics/storage-buffer.js';
import { BindUniformBufferFormat, BindStorageBufferFormat, BindTextureFormat, BindGroupFormat } from '../../platform/graphics/bind-group-format.js';
import { UniformBufferFormat, UniformFormat } from '../../platform/graphics/uniform-buffer-format.js';
import { UNIFORMTYPE_UINT, SHADERSTAGE_COMPUTE, SAMPLETYPE_UINT, SHADERLANGUAGE_WGSL, BUFFERUSAGE_COPY_SRC, BUFFERUSAGE_COPY_DST } from '../../platform/graphics/constants.js';
import { computeGsplatIntervalCullSource } from '../shader-lib/wgsl/chunks/gsplat/compute-gsplat-interval-cull.js';
import { computeGsplatIntervalScatterSource } from '../shader-lib/wgsl/chunks/gsplat/compute-gsplat-interval-scatter.js';
import { computeGsplatWriteIndirectArgsSource } from '../shader-lib/wgsl/chunks/gsplat/compute-gsplat-write-indirect-args.js';
import { PrefixSumKernel } from '../graphics/prefix-sum-kernel.js';
import { GSplatResourceBase } from '../gsplat/gsplat-resource-base.js';

const WORKGROUP_SIZE = 256;
const INDEX_COUNT = 6 * GSplatResourceBase.instanceSize;
const SORT_THREADS_PER_WORKGROUP = 256;
const INTERVAL_STRIDE = 4;
class GSplatIntervalCompaction {
		destroy() {
				this.compactedSplatIds?.destroy();
				this.intervalsBuffer?.destroy();
				this.countBuffer?.destroy();
				this.prefixSumKernel?.destroy();
				this.numSplatsBuffer?.destroy();
				this.sortElementCountBuffer?.destroy();
				this._destroyCullPass();
				this._scatterCompute?.shader?.destroy();
				this._scatterBindGroupFormat?.destroy();
				this._writeIndirectArgsCompute?.shader?.destroy();
				this._writeArgsBindGroupFormat?.destroy();
				this.compactedSplatIds = null;
				this.intervalsBuffer = null;
				this.countBuffer = null;
				this.prefixSumKernel = null;
				this.numSplatsBuffer = null;
				this.sortElementCountBuffer = null;
				this._scatterCompute = null;
				this._scatterBindGroupFormat = null;
				this._writeIndirectArgsCompute = null;
				this._writeArgsBindGroupFormat = null;
				this._cullUniformBufferFormat = null;
				this._scatterUniformBufferFormat = null;
				this._writeArgsUniformBufferFormat = null;
		}
		_destroyCullPass() {
				this._cullCompute?.shader?.destroy();
				this._cullBindGroupFormat?.destroy();
				this._cullCompute = null;
				this._cullBindGroupFormat = null;
		}
		_createUniformBufferFormats() {
				const device = this.device;
				this._cullUniformBufferFormat = new UniformBufferFormat(device, [
						new UniformFormat('numIntervals', UNIFORMTYPE_UINT),
						new UniformFormat('visWidth', UNIFORMTYPE_UINT)
				]);
				this._scatterUniformBufferFormat = new UniformBufferFormat(device, [
						new UniformFormat('numIntervals', UNIFORMTYPE_UINT),
						new UniformFormat('pad0', UNIFORMTYPE_UINT),
						new UniformFormat('pad1', UNIFORMTYPE_UINT),
						new UniformFormat('pad2', UNIFORMTYPE_UINT)
				]);
				this._writeArgsUniformBufferFormat = new UniformBufferFormat(device, [
						new UniformFormat('drawSlot', UNIFORMTYPE_UINT),
						new UniformFormat('indexCount', UNIFORMTYPE_UINT),
						new UniformFormat('dispatchSlotOffset', UNIFORMTYPE_UINT),
						new UniformFormat('totalSplats', UNIFORMTYPE_UINT)
				]);
		}
		_ensureCullPass(cullingEnabled) {
				if (this._cullCompute && cullingEnabled === this._cullingEnabled) {
						return;
				}
				this._destroyCullPass();
				this._cullingEnabled = cullingEnabled;
				const device = this.device;
				const suffix = cullingEnabled ? 'Culled' : '';
				const entries = [
						new BindUniformBufferFormat('uniforms', SHADERSTAGE_COMPUTE),
						new BindStorageBufferFormat('intervals', SHADERSTAGE_COMPUTE, true),
						new BindStorageBufferFormat('countBuffer', SHADERSTAGE_COMPUTE, false)
				];
				if (cullingEnabled) {
						entries.push(new BindTextureFormat('nodeVisibilityTexture', SHADERSTAGE_COMPUTE, undefined, SAMPLETYPE_UINT, false));
				}
				this._cullBindGroupFormat = new BindGroupFormat(device, entries);
				const cdefines = new Map([
						[
								'{WORKGROUP_SIZE}',
								WORKGROUP_SIZE.toString()
						]
				]);
				if (cullingEnabled) {
						cdefines.set('CULLING_ENABLED', '');
				}
				const shader = new Shader(device, {
						name: `GSplatIntervalCull${suffix}`,
						shaderLanguage: SHADERLANGUAGE_WGSL,
						cshader: computeGsplatIntervalCullSource,
						cdefines: cdefines,
						computeBindGroupFormat: this._cullBindGroupFormat,
						computeUniformBufferFormats: {
								uniforms: this._cullUniformBufferFormat
						}
				});
				this._cullCompute = new Compute(device, shader, `GSplatIntervalCull${suffix}`);
		}
		_createScatterCompute() {
				const device = this.device;
				this._scatterBindGroupFormat = new BindGroupFormat(device, [
						new BindUniformBufferFormat('uniforms', SHADERSTAGE_COMPUTE),
						new BindStorageBufferFormat('intervals', SHADERSTAGE_COMPUTE, true),
						new BindStorageBufferFormat('prefixSumBuffer', SHADERSTAGE_COMPUTE, true),
						new BindStorageBufferFormat('compactedOutput', SHADERSTAGE_COMPUTE, false)
				]);
				const cdefines = new Map([
						[
								'{WORKGROUP_SIZE}',
								WORKGROUP_SIZE.toString()
						]
				]);
				const shader = new Shader(device, {
						name: 'GSplatIntervalScatter',
						shaderLanguage: SHADERLANGUAGE_WGSL,
						cshader: computeGsplatIntervalScatterSource,
						cdefines: cdefines,
						computeBindGroupFormat: this._scatterBindGroupFormat,
						computeUniformBufferFormats: {
								uniforms: this._scatterUniformBufferFormat
						}
				});
				this._scatterCompute = new Compute(device, shader, 'GSplatIntervalScatter');
		}
		_createWriteIndirectArgsCompute() {
				const device = this.device;
				this._writeArgsBindGroupFormat = new BindGroupFormat(device, [
						new BindStorageBufferFormat('prefixSumBuffer', SHADERSTAGE_COMPUTE, true),
						new BindStorageBufferFormat('indirectDrawArgs', SHADERSTAGE_COMPUTE, false),
						new BindStorageBufferFormat('numSplatsBuf', SHADERSTAGE_COMPUTE, false),
						new BindStorageBufferFormat('indirectDispatchArgs', SHADERSTAGE_COMPUTE, false),
						new BindStorageBufferFormat('sortElementCountBuf', SHADERSTAGE_COMPUTE, false),
						new BindUniformBufferFormat('uniforms', SHADERSTAGE_COMPUTE)
				]);
				const cdefines = new Map([
						[
								'{INSTANCE_SIZE}',
								GSplatResourceBase.instanceSize
						],
						[
								'{SORT_THREADS_PER_WORKGROUP}',
								SORT_THREADS_PER_WORKGROUP
						]
				]);
				const shader = new Shader(device, {
						name: 'GSplatIntervalWriteIndirectArgs',
						shaderLanguage: SHADERLANGUAGE_WGSL,
						cshader: computeGsplatWriteIndirectArgsSource,
						cdefines: cdefines,
						computeBindGroupFormat: this._writeArgsBindGroupFormat,
						computeUniformBufferFormats: {
								uniforms: this._writeArgsUniformBufferFormat
						}
				});
				this._writeIndirectArgsCompute = new Compute(device, shader, 'GSplatIntervalWriteIndirectArgs');
		}
		_ensureCapacity(numIntervals, totalActiveSplats) {
				if (totalActiveSplats > this.allocatedCompactedCount) {
						this.compactedSplatIds?.destroy();
						this.allocatedCompactedCount = totalActiveSplats;
						this.compactedSplatIds = new StorageBuffer(this.device, totalActiveSplats * 4, BUFFERUSAGE_COPY_SRC);
				}
				const requiredCountSize = numIntervals + 1;
				if (requiredCountSize > this.allocatedCountBufferSize) {
						this.countBuffer?.destroy();
						this.allocatedCountBufferSize = requiredCountSize;
						this.countBuffer = new StorageBuffer(this.device, requiredCountSize * 4);
						if (this.prefixSumKernel) {
								this.prefixSumKernel.destroyPasses();
						}
				}
		}
		uploadIntervals(worldState) {
				if (worldState.version === this._uploadedVersion) return;
				this._uploadedVersion = worldState.version;
				const splats = worldState.splats;
				const numIntervals = worldState.totalIntervals;
				if (numIntervals === 0) return;
				if (numIntervals > this.allocatedIntervalCount) {
						this.intervalsBuffer?.destroy();
						this.allocatedIntervalCount = numIntervals;
						this.intervalsBuffer = new StorageBuffer(this.device, numIntervals * INTERVAL_STRIDE * 4, BUFFERUSAGE_COPY_DST);
				}
				const data = new Uint32Array(numIntervals * INTERVAL_STRIDE);
				let writeIdx = 0;
				for(let s = 0; s < splats.length; s++){
						const splat = splats[s];
						if (splat.intervals.length > 0) {
								const nodeIndices = splat.intervalNodeIndices;
								for(let i = 0; i < splat.intervals.length; i += 2){
										const count = splat.intervals[i + 1] - splat.intervals[i];
										data[writeIdx++] = splat.intervalOffsets[i / 2];
										data[writeIdx++] = count;
										data[writeIdx++] = splat.boundsBaseIndex + (nodeIndices.length > 0 ? nodeIndices[i / 2] : 0);
										data[writeIdx++] = 0;
								}
						} else {
								data[writeIdx++] = splat.intervalOffsets[0];
								data[writeIdx++] = splat.activeSplats;
								data[writeIdx++] = splat.boundsBaseIndex;
								data[writeIdx++] = 0;
						}
				}
				this.intervalsBuffer.write(0, data, 0, numIntervals * INTERVAL_STRIDE);
		}
		dispatchCompact(nodeVisibilityTexture, numIntervals, totalActiveSplats, cullingEnabled) {
				if (numIntervals === 0) return;
				this._ensureCapacity(numIntervals, totalActiveSplats);
				this._ensureCullPass(cullingEnabled);
				const cullCompute = this._cullCompute;
				cullCompute.setParameter('intervals', this.intervalsBuffer);
				cullCompute.setParameter('countBuffer', this.countBuffer);
				if (cullingEnabled) {
						cullCompute.setParameter('nodeVisibilityTexture', nodeVisibilityTexture);
				}
				cullCompute.setParameter('numIntervals', numIntervals);
				cullCompute.setParameter('visWidth', cullingEnabled ? nodeVisibilityTexture.width : 0);
				const cullWorkgroups = Math.ceil(numIntervals / WORKGROUP_SIZE);
				cullCompute.setupDispatch(cullWorkgroups);
				this.device.computeDispatch([
						cullCompute
				], 'GSplatIntervalCull');
				const prefixCount = numIntervals + 1;
				this.prefixSumKernel.resize(this.countBuffer, prefixCount);
				this.prefixSumKernel.dispatch(this.device);
				const scatterCompute = this._scatterCompute;
				scatterCompute.setParameter('intervals', this.intervalsBuffer);
				scatterCompute.setParameter('prefixSumBuffer', this.countBuffer);
				scatterCompute.setParameter('compactedOutput', this.compactedSplatIds);
				scatterCompute.setParameter('numIntervals', numIntervals);
				scatterCompute.setParameter('pad0', 0);
				scatterCompute.setParameter('pad1', 0);
				scatterCompute.setParameter('pad2', 0);
				scatterCompute.setupDispatch(numIntervals);
				this.device.computeDispatch([
						scatterCompute
				], 'GSplatIntervalScatter');
		}
		writeIndirectArgs(drawSlot, dispatchSlot, numIntervals) {
				const compute = this._writeIndirectArgsCompute;
				compute.setParameter('prefixSumBuffer', this.countBuffer);
				compute.setParameter('indirectDrawArgs', this.device.indirectDrawBuffer);
				compute.setParameter('numSplatsBuf', this.numSplatsBuffer);
				compute.setParameter('indirectDispatchArgs', this.device.indirectDispatchBuffer);
				compute.setParameter('sortElementCountBuf', this.sortElementCountBuffer);
				compute.setParameter('drawSlot', drawSlot);
				compute.setParameter('indexCount', INDEX_COUNT);
				compute.setParameter('dispatchSlotOffset', dispatchSlot * 3);
				compute.setParameter('totalSplats', numIntervals);
				compute.setupDispatch(1);
				this.device.computeDispatch([
						compute
				], 'GSplatIntervalWriteIndirectArgs');
		}
		constructor(device){
				this.compactedSplatIds = null;
				this.intervalsBuffer = null;
				this.countBuffer = null;
				this.prefixSumKernel = null;
				this.numSplatsBuffer = null;
				this.sortElementCountBuffer = null;
				this.allocatedCompactedCount = 0;
				this.allocatedIntervalCount = 0;
				this.allocatedCountBufferSize = 0;
				this._uploadedVersion = -1;
				this._cullingEnabled = false;
				this._cullCompute = null;
				this._scatterCompute = null;
				this._writeIndirectArgsCompute = null;
				this._cullBindGroupFormat = null;
				this._scatterBindGroupFormat = null;
				this._writeArgsBindGroupFormat = null;
				this._cullUniformBufferFormat = null;
				this._scatterUniformBufferFormat = null;
				this._writeArgsUniformBufferFormat = null;
				this.device = device;
				this.numSplatsBuffer = new StorageBuffer(device, 4, BUFFERUSAGE_COPY_SRC | BUFFERUSAGE_COPY_DST);
				this.sortElementCountBuffer = new StorageBuffer(device, 4, BUFFERUSAGE_COPY_SRC | BUFFERUSAGE_COPY_DST);
				this.prefixSumKernel = new PrefixSumKernel(device);
				this._createUniformBufferFormats();
				this._createScatterCompute();
				this._createWriteIndirectArgsCompute();
		}
}

export { GSplatIntervalCompaction };
