// playcanvas/build/playcanvas.prf/src/scene/gsplat-unified/gsplat-work-buffer.js
import { Frustum } from '../../core/shape/frustum.js';
import { Mat4 } from '../../core/math/mat4.js';
import { Vec2 } from '../../core/math/vec2.js';
import { PIXELFORMAT_RGBA16U, getGlslShaderType, SEMANTIC_POSITION, BUFFERUSAGE_COPY_DST, PIXELFORMAT_RGBA32F, PIXELFORMAT_R32U, ADDRESS_CLAMP_TO_EDGE } from '../../platform/graphics/constants.js';
import { RenderTarget } from '../../platform/graphics/render-target.js';
import { StorageBuffer } from '../../platform/graphics/storage-buffer.js';
import { Texture } from '../../platform/graphics/texture.js';
import { TextureUtils } from '../../platform/graphics/texture-utils.js';
import { UploadStream } from '../../platform/graphics/upload-stream.js';
import { QuadRender } from '../graphics/quad-render.js';
import { ShaderUtils } from '../shader-lib/shader-utils.js';
import glslGsplatCopyToWorkBufferPS from '../shader-lib/glsl/chunks/gsplat/frag/gsplatCopyToWorkbuffer.js';
import wgslGsplatCopyToWorkBufferPS from '../shader-lib/wgsl/chunks/gsplat/frag/gsplatCopyToWorkbuffer.js';
import glslGsplatCopyInstancedQuadVS from '../shader-lib/glsl/chunks/gsplat/vert/gsplatCopyInstancedQuad.js';
import wgslGsplatCopyInstancedQuadVS from '../shader-lib/wgsl/chunks/gsplat/vert/gsplatCopyInstancedQuad.js';
import { GSplatNodeCullRenderPass } from './gsplat-node-cull-render-pass.js';
import { GSplatWorkBufferRenderPass } from './gsplat-work-buffer-render-pass.js';
import { GSplatStreams } from '../gsplat/gsplat-streams.js';

let id = 0;
const tmpSize = new Vec2();
const _viewProjMat = new Mat4();
const _frustum = new Frustum();
const _frustumPlanes = new Float32Array(24);
class WorkBufferRenderInfo {
		destroy() {
				this.material?.destroy();
				this.quadRender?.destroy();
		}
		constructor(device, key, material, colorOnly, format){
				this.material = material;
				const clonedDefines = new Map(material.defines);
				const colorStream = format.getStream('dataColor');
				if (colorStream.format === PIXELFORMAT_RGBA16U) {
						clonedDefines.set('GSPLAT_COLOR_UINT', '');
				}
				if (colorOnly) {
						clonedDefines.set('GSPLAT_COLOR_ONLY', '');
				}
				if (format.getStream('pcId')) {
						clonedDefines.set('GSPLAT_ID', '');
				}
				if (format.getStream('pcNodeIndex')) {
						clonedDefines.set('GSPLAT_NODE_INDEX', '');
				}
				const fragmentIncludes = material.hasShaderChunks ? device.isWebGPU ? material.shaderChunks.wgsl : material.shaderChunks.glsl : undefined;
				const outputStreams = colorOnly ? [
						colorStream
				] : [
						...format.streams,
						...format.extraStreams
				];
				const fragmentOutputTypes = [];
				for (const stream of outputStreams){
						const info = getGlslShaderType(stream.format);
						fragmentOutputTypes.push(info.returnType);
				}
				const useInstanced = clonedDefines.has('GSPLAT_LOD');
				const shaderOptions = {
						uniqueName: `SplatCopyToWorkBuffer:${key}`,
						attributes: {
								vertex_position: SEMANTIC_POSITION
						},
						vertexDefines: clonedDefines,
						fragmentDefines: clonedDefines,
						fragmentGLSL: glslGsplatCopyToWorkBufferPS,
						fragmentWGSL: wgslGsplatCopyToWorkBufferPS,
						fragmentIncludes: fragmentIncludes,
						fragmentOutputTypes: fragmentOutputTypes
				};
				if (useInstanced) {
						shaderOptions.vertexGLSL = glslGsplatCopyInstancedQuadVS;
						shaderOptions.vertexWGSL = wgslGsplatCopyInstancedQuadVS;
				} else {
						shaderOptions.vertexChunk = 'fullscreenQuadVS';
				}
				const shader = ShaderUtils.createShader(device, shaderOptions);
				this.quadRender = new QuadRender(shader);
		}
}
class GSplatWorkBuffer {
		_createRenderTargets() {
				this.renderTarget?.destroy();
				this.colorRenderTarget?.destroy();
				const colorBuffers = this.streams.getTexturesInOrder();
				this.renderTarget = new RenderTarget({
						name: `GsplatWorkBuffer-MRT-${this.id}`,
						colorBuffers: colorBuffers,
						depth: false,
						flipY: true
				});
				const colorTexture = this.streams.getTexture('dataColor');
				this.colorRenderTarget = new RenderTarget({
						name: `GsplatWorkBuffer-Color-${this.id}`,
						colorBuffer: colorTexture,
						depth: false,
						flipY: true
				});
				this.renderPass?.init(this.renderTarget);
				this.colorRenderPass?.init(this.colorRenderTarget);
		}
		syncWithFormat() {
				const prevVersion = this.streams._formatVersion;
				this.streams.syncWithFormat(this.format);
				if (prevVersion !== this.streams._formatVersion) {
						this._createRenderTargets();
				}
		}
		getTexture(name) {
				return this.streams.getTexture(name);
		}
		destroy() {
				this.renderPass?.destroy();
				this.colorRenderPass?.destroy();
				this.streams.destroy();
				this.orderTexture?.destroy();
				this.orderBuffer?.destroy();
				this.renderTarget?.destroy();
				this.colorRenderTarget?.destroy();
				this.uploadStream.destroy();
				this.boundsSphereTexture?.destroy();
				this.boundsTransformIndexTexture?.destroy();
				this.nodeVisibilityTexture?.destroy();
				this.cullingRenderTarget?.destroy();
				this.cullingPass?.destroy();
				this.transformsTexture?.destroy();
		}
		get textureSize() {
				return this.streams.textureDimensions.x;
		}
		setOrderData(data) {
				this.textureSize;
				if (this.device.isWebGPU) {
						this.uploadStream.upload(data, this.orderBuffer, 0, data.length);
				} else {
						this.uploadStream.upload(data, this.orderTexture, 0, data.length);
				}
		}
		resize(textureSize) {
				this.renderTarget.resize(textureSize, textureSize);
				this.colorRenderTarget.resize(textureSize, textureSize);
				this.streams.resize(textureSize, textureSize);
				if (this.device.isWebGPU) {
						const newByteSize = textureSize * textureSize * 4;
						if (this.orderBuffer.byteSize < newByteSize) {
								this.orderBuffer.destroy();
								this.orderBuffer = new StorageBuffer(this.device, newByteSize, BUFFERUSAGE_COPY_DST);
						}
				} else {
						this.orderTexture.resize(textureSize, textureSize);
				}
		}
		render(splats, cameraNode, colorsByLod, changedAllocIds = null) {
				if (this.renderPass.update(splats, cameraNode, colorsByLod, changedAllocIds)) {
						this.renderPass.render();
				}
		}
		renderColor(splats, cameraNode, colorsByLod) {
				if (this.colorRenderPass.update(splats, cameraNode, colorsByLod)) {
						this.colorRenderPass.render();
				}
		}
		updateBoundsTexture(boundsGroups) {
				let totalEntries = 0;
				for(let i = 0; i < boundsGroups.length; i++){
						totalEntries += boundsGroups[i].numBoundsEntries;
				}
				this.totalBoundsEntries = totalEntries;
				if (totalEntries === 0) return;
				const { x: width, y: height } = TextureUtils.calcTextureSize(totalEntries, tmpSize, 32);
				if (!this.boundsSphereTexture) {
						this.boundsSphereTexture = Texture.createDataTexture2D(this.device, 'boundsSphereTexture', width, height, PIXELFORMAT_RGBA32F);
				} else {
						this.boundsSphereTexture.resize(width, height);
				}
				if (!this.boundsTransformIndexTexture) {
						this.boundsTransformIndexTexture = Texture.createDataTexture2D(this.device, 'boundsTransformIndexTexture', width, height, PIXELFORMAT_R32U);
				} else {
						this.boundsTransformIndexTexture.resize(width, height);
				}
				const sphereData = this.boundsSphereTexture.lock();
				const indexData = this.boundsTransformIndexTexture.lock();
				for(let i = 0; i < boundsGroups.length; i++){
						const group = boundsGroups[i];
						const base = group.boundsBaseIndex;
						const count = group.numBoundsEntries;
						group.splat.writeBoundsSpheres(sphereData, base * 4);
						for(let j = 0; j < count; j++){
								indexData[base + j] = i;
						}
				}
				this.boundsSphereTexture.unlock();
				this.boundsTransformIndexTexture.unlock();
		}
		updateTransformsTexture(boundsGroups) {
				const numMatrices = boundsGroups.length;
				if (numMatrices === 0) return;
				const totalTexels = numMatrices * 3;
				const { x: width, y: height } = TextureUtils.calcTextureSize(totalTexels, tmpSize, 3);
				if (!this.transformsTexture) {
						this.transformsTexture = Texture.createDataTexture2D(this.device, 'transformsTexture', width, height, PIXELFORMAT_RGBA32F);
				} else {
						this.transformsTexture.resize(width, height);
				}
				const data = this.transformsTexture.lock();
				let offset = 0;
				for(let i = 0; i < boundsGroups.length; i++){
						const m = boundsGroups[i].splat.node.getWorldTransform().data;
						data[offset++] = m[0];
						data[offset++] = m[4];
						data[offset++] = m[8];
						data[offset++] = m[12];
						data[offset++] = m[1];
						data[offset++] = m[5];
						data[offset++] = m[9];
						data[offset++] = m[13];
						data[offset++] = m[2];
						data[offset++] = m[6];
						data[offset++] = m[10];
						data[offset++] = m[14];
				}
				this.transformsTexture.unlock();
		}
		updateNodeVisibility(projectionMatrix, viewMatrix) {
				if (this.totalBoundsEntries === 0 || !this.boundsSphereTexture || !this.boundsTransformIndexTexture || !this.transformsTexture) {
						return;
				}
				_viewProjMat.mul2(projectionMatrix, viewMatrix);
				_frustum.setFromMat4(_viewProjMat);
				for(let p = 0; p < 6; p++){
						const plane = _frustum.planes[p];
						_frustumPlanes[p * 4 + 0] = plane.normal.x;
						_frustumPlanes[p * 4 + 1] = plane.normal.y;
						_frustumPlanes[p * 4 + 2] = plane.normal.z;
						_frustumPlanes[p * 4 + 3] = plane.distance;
				}
				const width = this.boundsSphereTexture.width / 32;
				const height = this.boundsSphereTexture.height;
				if (!this.nodeVisibilityTexture) {
						this.nodeVisibilityTexture = Texture.createDataTexture2D(this.device, 'nodeVisibilityTexture', width, height, PIXELFORMAT_R32U);
						this.cullingRenderTarget = new RenderTarget({
								name: 'NodeCullingRT',
								colorBuffer: this.nodeVisibilityTexture,
								depth: false
						});
				} else if (this.nodeVisibilityTexture.width !== width || this.nodeVisibilityTexture.height !== height) {
						this.nodeVisibilityTexture.resize(width, height);
						this.cullingRenderTarget.resize(width, height);
				}
				if (!this.cullingPass) {
						this.cullingPass = new GSplatNodeCullRenderPass(this.device);
						this.cullingPass.init(this.cullingRenderTarget);
						this.cullingPass.colorOps.clear = true;
						this.cullingPass.colorOps.clearValue.set(0, 0, 0, 0);
				}
				this.cullingPass.setup(this.boundsSphereTexture, this.boundsTransformIndexTexture, this.transformsTexture, this.totalBoundsEntries, _frustumPlanes);
				this.cullingPass.render();
		}
		constructor(device, format){
				this.id = id++;
				this.boundsSphereTexture = null;
				this.boundsTransformIndexTexture = null;
				this.nodeVisibilityTexture = null;
				this.cullingRenderTarget = null;
				this.cullingPass = null;
				this.totalBoundsEntries = 0;
				this.transformsTexture = null;
				this.device = device;
				this.format = format;
				this.streams = new GSplatStreams(device);
				this.streams.init(format, 1);
				this._createRenderTargets();
				this.uploadStream = new UploadStream(device);
				if (device.isWebGPU) {
						this.orderBuffer = new StorageBuffer(device, 4, BUFFERUSAGE_COPY_DST);
				} else {
						this.orderTexture = new Texture(device, {
								name: 'SplatGlobalOrder',
								width: 1,
								height: 1,
								format: PIXELFORMAT_R32U,
								mipmaps: false,
								addressU: ADDRESS_CLAMP_TO_EDGE,
								addressV: ADDRESS_CLAMP_TO_EDGE
						});
				}
				this.renderPass = new GSplatWorkBufferRenderPass(device, this);
				this.renderPass.init(this.renderTarget);
				this.colorRenderPass = new GSplatWorkBufferRenderPass(device, this, true);
				this.colorRenderPass.init(this.colorRenderTarget);
		}
}

export { GSplatWorkBuffer, WorkBufferRenderInfo };
