// playcanvas/build/playcanvas.prf/src/scene/gsplat-unified/gsplat-compaction.js
import { Vec2 } from '../../core/math/vec2.js';
import { Compute } from '../../platform/graphics/compute.js';
import { Shader } from '../../platform/graphics/shader.js';
import { StorageBuffer } from '../../platform/graphics/storage-buffer.js';
import { BindUniformBufferFormat, BindTextureFormat, BindStorageBufferFormat, BindGroupFormat } from '../../platform/graphics/bind-group-format.js';
import { UniformBufferFormat, UniformFormat } from '../../platform/graphics/uniform-buffer-format.js';
import { UNIFORMTYPE_UINT, SHADERSTAGE_COMPUTE, SAMPLETYPE_UINT, SHADERLANGUAGE_WGSL, BUFFERUSAGE_COPY_SRC, BUFFERUSAGE_COPY_DST } from '../../platform/graphics/constants.js';
import { computeGsplatCompactFlagSource } from '../shader-lib/wgsl/chunks/gsplat/compute-gsplat-compact-flag.js';
import { computeGsplatCompactScatterSource } from '../shader-lib/wgsl/chunks/gsplat/compute-gsplat-compact-scatter.js';
import { computeGsplatWriteIndirectArgsSource } from '../shader-lib/wgsl/chunks/gsplat/compute-gsplat-write-indirect-args.js';
import { PrefixSumKernel } from '../graphics/prefix-sum-kernel.js';
import { GSplatResourceBase } from '../gsplat/gsplat-resource-base.js';

const WORKGROUP_SIZE = 256;
const SPLATS_PER_THREAD = 4;
const INDEX_COUNT = 6 * GSplatResourceBase.instanceSize;
const SORT_THREADS_PER_WORKGROUP = 256;
const _dispatchSize = new Vec2();
class GSplatCompaction {
		destroy() {
				this.compactedSplatIds?.destroy();
				this.flagBuffer?.destroy();
				this.prefixSumKernel?.destroy();
				this.numSplatsBuffer?.destroy();
				this.sortElementCountBuffer?.destroy();
				this._destroyCompactPasses();
				this._writeIndirectArgsCompute?.shader?.destroy();
				this._writeArgsBindGroupFormat?.destroy();
				this.compactedSplatIds = null;
				this.flagBuffer = null;
				this.prefixSumKernel = null;
				this.numSplatsBuffer = null;
				this.sortElementCountBuffer = null;
				this._writeIndirectArgsCompute = null;
				this._writeArgsBindGroupFormat = null;
				this._flagUniformBufferFormat = null;
				this._scatterUniformBufferFormat = null;
				this._writeArgsUniformBufferFormat = null;
		}
		_destroyCompactPasses() {
				this._flagCompute?.shader?.destroy();
				this._scatterCompute?.shader?.destroy();
				this._flagBindGroupFormat?.destroy();
				this._scatterBindGroupFormat?.destroy();
				this._flagCompute = null;
				this._scatterCompute = null;
				this._flagBindGroupFormat = null;
				this._scatterBindGroupFormat = null;
		}
		_createUniformBufferFormats() {
				const device = this.device;
				this._flagUniformBufferFormat = new UniformBufferFormat(device, [
						new UniformFormat('totalSplats', UNIFORMTYPE_UINT),
						new UniformFormat('textureWidth', UNIFORMTYPE_UINT),
						new UniformFormat('visWidth', UNIFORMTYPE_UINT),
						new UniformFormat('totalThreads', UNIFORMTYPE_UINT),
						new UniformFormat('numWorkgroupsX', UNIFORMTYPE_UINT)
				]);
				this._scatterUniformBufferFormat = new UniformBufferFormat(device, [
						new UniformFormat('totalSplats', UNIFORMTYPE_UINT),
						new UniformFormat('numWorkgroupsX', UNIFORMTYPE_UINT),
						new UniformFormat('pad1', UNIFORMTYPE_UINT),
						new UniformFormat('pad2', UNIFORMTYPE_UINT)
				]);
				this._writeArgsUniformBufferFormat = new UniformBufferFormat(device, [
						new UniformFormat('drawSlot', UNIFORMTYPE_UINT),
						new UniformFormat('indexCount', UNIFORMTYPE_UINT),
						new UniformFormat('dispatchSlotOffset', UNIFORMTYPE_UINT),
						new UniformFormat('totalSplats', UNIFORMTYPE_UINT)
				]);
		}
		_ensureCompactPasses(useSortedOrder) {
				if (this._flagCompute && useSortedOrder === this._useSortedOrder) {
						return;
				}
				this._destroyCompactPasses();
				this._useSortedOrder = useSortedOrder;
				const device = this.device;
				const suffix = useSortedOrder ? 'Sorted' : '';
				const flagEntries = [
						new BindUniformBufferFormat('uniforms', SHADERSTAGE_COMPUTE),
						new BindTextureFormat('pcNodeIndex', SHADERSTAGE_COMPUTE, undefined, SAMPLETYPE_UINT, false),
						new BindTextureFormat('nodeVisibilityTexture', SHADERSTAGE_COMPUTE, undefined, SAMPLETYPE_UINT, false),
						new BindStorageBufferFormat('flagBuffer', SHADERSTAGE_COMPUTE, false)
				];
				if (useSortedOrder) {
						flagEntries.push(new BindStorageBufferFormat('sortedOrder', SHADERSTAGE_COMPUTE, true));
				}
				this._flagBindGroupFormat = new BindGroupFormat(device, flagEntries);
				const scatterEntries = [
						new BindUniformBufferFormat('uniforms', SHADERSTAGE_COMPUTE),
						new BindStorageBufferFormat('prefixSumBuffer', SHADERSTAGE_COMPUTE, true),
						new BindStorageBufferFormat('compactedOutput', SHADERSTAGE_COMPUTE, false)
				];
				if (useSortedOrder) {
						scatterEntries.push(new BindStorageBufferFormat('sortedOrder', SHADERSTAGE_COMPUTE, true));
				}
				this._scatterBindGroupFormat = new BindGroupFormat(device, scatterEntries);
				this._flagCompute = this._createCompactShader(`GSplatCompactFlag${suffix}`, computeGsplatCompactFlagSource, this._flagBindGroupFormat, this._flagUniformBufferFormat, useSortedOrder);
				this._scatterCompute = this._createCompactShader(`GSplatCompactScatter${suffix}`, computeGsplatCompactScatterSource, this._scatterBindGroupFormat, this._scatterUniformBufferFormat, useSortedOrder);
		}
		_createCompactShader(name, source, bindGroupFormat, uniformBufferFormat, useSortedOrder) {
				const cdefines = new Map([
						[
								'{WORKGROUP_SIZE}',
								WORKGROUP_SIZE
						]
				]);
				if (useSortedOrder) {
						cdefines.set('USE_SORTED_ORDER', '');
				}
				const shader = new Shader(this.device, {
						name: name,
						shaderLanguage: SHADERLANGUAGE_WGSL,
						cshader: source,
						cdefines: cdefines,
						computeBindGroupFormat: bindGroupFormat,
						computeUniformBufferFormats: {
								uniforms: uniformBufferFormat
						}
				});
				return new Compute(this.device, shader, name);
		}
		_createWriteIndirectArgsCompute() {
				const device = this.device;
				this._writeArgsBindGroupFormat = new BindGroupFormat(device, [
						new BindStorageBufferFormat('prefixSumBuffer', SHADERSTAGE_COMPUTE, true),
						new BindStorageBufferFormat('indirectDrawArgs', SHADERSTAGE_COMPUTE, false),
						new BindStorageBufferFormat('numSplatsBuf', SHADERSTAGE_COMPUTE, false),
						new BindStorageBufferFormat('indirectDispatchArgs', SHADERSTAGE_COMPUTE, false),
						new BindStorageBufferFormat('sortElementCountBuf', SHADERSTAGE_COMPUTE, false),
						new BindUniformBufferFormat('uniforms', SHADERSTAGE_COMPUTE)
				]);
				const cdefines = new Map([
						[
								'{INSTANCE_SIZE}',
								GSplatResourceBase.instanceSize
						],
						[
								'{SORT_THREADS_PER_WORKGROUP}',
								SORT_THREADS_PER_WORKGROUP
						]
				]);
				const shader = new Shader(device, {
						name: 'GSplatWriteIndirectArgs',
						shaderLanguage: SHADERLANGUAGE_WGSL,
						cshader: computeGsplatWriteIndirectArgsSource,
						cdefines: cdefines,
						computeBindGroupFormat: this._writeArgsBindGroupFormat,
						computeUniformBufferFormats: {
								uniforms: this._writeArgsUniformBufferFormat
						}
				});
				this._writeIndirectArgsCompute = new Compute(device, shader, 'GSplatWriteIndirectArgs');
		}
		_ensureCapacity(totalSplats) {
				if (totalSplats > this.allocatedCount) {
						this.compactedSplatIds?.destroy();
						this.allocatedCount = totalSplats;
						this.compactedSplatIds = new StorageBuffer(this.device, totalSplats * 4, BUFFERUSAGE_COPY_SRC);
				}
				const requiredFlagCount = totalSplats + 1;
				if (requiredFlagCount > this.allocatedFlagCount) {
						this.flagBuffer?.destroy();
						this.allocatedFlagCount = requiredFlagCount;
						this.flagBuffer = new StorageBuffer(this.device, requiredFlagCount * 4);
						if (this.prefixSumKernel) {
								this.prefixSumKernel.destroyPasses();
						}
				}
		}
		dispatchCompact(pcNodeIndexTexture, nodeVisibilityTexture, totalSplats, textureWidth, sortedOrderBuffer) {
				this._ensureCapacity(totalSplats);
				const useSortedOrder = !!sortedOrderBuffer;
				this._ensureCompactPasses(useSortedOrder);
				const flagCompute = this._flagCompute;
				flagCompute.setParameter('pcNodeIndex', pcNodeIndexTexture);
				flagCompute.setParameter('nodeVisibilityTexture', nodeVisibilityTexture);
				flagCompute.setParameter('flagBuffer', this.flagBuffer);
				if (useSortedOrder) {
						flagCompute.setParameter('sortedOrder', sortedOrderBuffer);
				}
				const flagWorkgroups = Math.ceil(totalSplats / (WORKGROUP_SIZE * SPLATS_PER_THREAD));
				Compute.calcDispatchSize(flagWorkgroups, _dispatchSize);
				const totalThreads = _dispatchSize.x * _dispatchSize.y * WORKGROUP_SIZE;
				flagCompute.setParameter('totalSplats', totalSplats);
				flagCompute.setParameter('textureWidth', textureWidth);
				flagCompute.setParameter('visWidth', nodeVisibilityTexture.width);
				flagCompute.setParameter('totalThreads', totalThreads);
				flagCompute.setParameter('numWorkgroupsX', _dispatchSize.x);
				flagCompute.setupDispatch(_dispatchSize.x, _dispatchSize.y, 1);
				this.device.computeDispatch([
						flagCompute
				], 'GSplatCompactFlag');
				const prefixCount = totalSplats + 1;
				this.prefixSumKernel.resize(this.flagBuffer, prefixCount);
				this.prefixSumKernel.dispatch(this.device);
				const scatterCompute = this._scatterCompute;
				scatterCompute.setParameter('prefixSumBuffer', this.flagBuffer);
				scatterCompute.setParameter('compactedOutput', this.compactedSplatIds);
				if (useSortedOrder) {
						scatterCompute.setParameter('sortedOrder', sortedOrderBuffer);
				}
				const scatterWorkgroups = Math.ceil(totalSplats / WORKGROUP_SIZE);
				Compute.calcDispatchSize(scatterWorkgroups, _dispatchSize);
				scatterCompute.setParameter('totalSplats', totalSplats);
				scatterCompute.setParameter('numWorkgroupsX', _dispatchSize.x);
				scatterCompute.setParameter('pad1', 0);
				scatterCompute.setParameter('pad2', 0);
				scatterCompute.setupDispatch(_dispatchSize.x, _dispatchSize.y, 1);
				this.device.computeDispatch([
						scatterCompute
				], 'GSplatCompactScatter');
		}
		writeIndirectArgs(drawSlot, dispatchSlot, totalSplats) {
				const compute = this._writeIndirectArgsCompute;
				compute.setParameter('prefixSumBuffer', this.flagBuffer);
				compute.setParameter('indirectDrawArgs', this.device.indirectDrawBuffer);
				compute.setParameter('numSplatsBuf', this.numSplatsBuffer);
				compute.setParameter('indirectDispatchArgs', this.device.indirectDispatchBuffer);
				compute.setParameter('sortElementCountBuf', this.sortElementCountBuffer);
				compute.setParameter('drawSlot', drawSlot);
				compute.setParameter('indexCount', INDEX_COUNT);
				compute.setParameter('dispatchSlotOffset', dispatchSlot * 3);
				compute.setParameter('totalSplats', totalSplats);
				compute.setupDispatch(1);
				this.device.computeDispatch([
						compute
				], 'GSplatWriteIndirectArgs');
		}
		constructor(device){
				this.compactedSplatIds = null;
				this.flagBuffer = null;
				this.prefixSumKernel = null;
				this.numSplatsBuffer = null;
				this.sortElementCountBuffer = null;
				this.allocatedCount = 0;
				this.allocatedFlagCount = 0;
				this._useSortedOrder = false;
				this._flagCompute = null;
				this._scatterCompute = null;
				this._writeIndirectArgsCompute = null;
				this._flagBindGroupFormat = null;
				this._scatterBindGroupFormat = null;
				this._writeArgsBindGroupFormat = null;
				this._flagUniformBufferFormat = null;
				this._scatterUniformBufferFormat = null;
				this._writeArgsUniformBufferFormat = null;
				this.device = device;
				this.numSplatsBuffer = new StorageBuffer(device, 4, BUFFERUSAGE_COPY_SRC | BUFFERUSAGE_COPY_DST);
				this.sortElementCountBuffer = new StorageBuffer(device, 4, BUFFERUSAGE_COPY_SRC | BUFFERUSAGE_COPY_DST);
				this.prefixSumKernel = new PrefixSumKernel(device);
				this._createUniformBufferFormats();
				this._createWriteIndirectArgsCompute();
		}
}

export { GSplatCompaction };
