// playcanvas/build/playcanvas/src/scene/gsplat-unified/gsplat-params.js
import { PIXELFORMAT_R32U, PIXELFORMAT_RGBA32U, PIXELFORMAT_RGBA16F, PIXELFORMAT_RGBA16U, PIXELFORMAT_RG32U } from '../../platform/graphics/constants.js';
import { ShaderMaterial } from '../materials/shader-material.js';
import { GSplatFormat } from '../gsplat/gsplat-format.js';
import { GSPLATDATA_COMPACT } from '../constants.js';
import glslCompactRead from '../shader-lib/glsl/chunks/gsplat/vert/formats/containerCompactRead.js';
import glslCompactWrite from '../shader-lib/glsl/chunks/gsplat/frag/formats/containerCompactWrite.js';
import glslPackedRead from '../shader-lib/glsl/chunks/gsplat/vert/formats/containerPackedRead.js';
import glslPackedWrite from '../shader-lib/glsl/chunks/gsplat/frag/formats/containerPackedWrite.js';
import wgslCompactRead from '../shader-lib/wgsl/chunks/gsplat/vert/formats/containerCompactRead.js';
import wgslCompactWrite from '../shader-lib/wgsl/chunks/gsplat/frag/formats/containerCompactWrite.js';
import wgslPackedRead from '../shader-lib/wgsl/chunks/gsplat/vert/formats/containerPackedRead.js';
import wgslPackedWrite from '../shader-lib/wgsl/chunks/gsplat/frag/formats/containerPackedWrite.js';

class GSplatParams {
		_createFormat(dataFormat) {
				let format;
				if (dataFormat === GSPLATDATA_COMPACT) {
						format = new GSplatFormat(this._device, [
								{
										name: 'dataColor',
										format: PIXELFORMAT_R32U
								},
								{
										name: 'dataTransformA',
										format: PIXELFORMAT_RGBA32U
								},
								{
										name: 'dataTransformB',
										format: PIXELFORMAT_R32U
								}
						], {
								readGLSL: glslCompactRead,
								readWGSL: wgslCompactRead
						});
						format.setWriteCode(glslCompactWrite, wgslCompactWrite);
				} else {
						const colorFormat = this._device.getRenderableHdrFormat([
								PIXELFORMAT_RGBA16F
						]) || PIXELFORMAT_RGBA16U;
						format = new GSplatFormat(this._device, [
								{
										name: 'dataColor',
										format: colorFormat
								},
								{
										name: 'dataTransformA',
										format: PIXELFORMAT_RGBA32U
								},
								{
										name: 'dataTransformB',
										format: PIXELFORMAT_RG32U
								}
						], {
								readGLSL: glslPackedRead,
								readWGSL: wgslPackedRead
						});
						format.setWriteCode(glslPackedWrite, wgslPackedWrite);
				}
				format.allowStreamRemoval = true;
				return format;
		}
		set gpuSorting(value) {
				if (value !== this._gpuSorting) {
						this._gpuSorting = value;
						this._syncNodeIndexStream();
				}
		}
		get gpuSorting() {
				return this._gpuSorting;
		}
		set colorizeLod(value) {
				if (this._colorizeLod !== value) {
						this._colorizeLod = value;
						this.dirty = true;
				}
		}
		get colorizeLod() {
				return this._colorizeLod;
		}
		set enableIds(value) {
				if (value && !this._enableIds) {
						this._enableIds = true;
						if (!this._format.getStream('pcId')) {
								this._format.addExtraStreams([
										{
												name: 'pcId',
												format: PIXELFORMAT_R32U
										}
								]);
						}
				} else if (!value && this._enableIds) {
						this._enableIds = false;
						this._format.removeExtraStreams([
								'pcId'
						]);
				}
		}
		get enableIds() {
				return this._enableIds;
		}
		set culling(value) {
				if (value !== this._culling) {
						this._culling = value;
						this._syncNodeIndexStream();
				}
		}
		get culling() {
				return this._culling;
		}
		_syncNodeIndexStream() {
				const needsNodeIndex = this._culling && !(this._gpuSorting && this._device.isWebGPU);
				const hasNodeIndex = !!this._format.getStream('pcNodeIndex');
				if (needsNodeIndex && !hasNodeIndex) {
						this._format.addExtraStreams([
								{
										name: 'pcNodeIndex',
										format: PIXELFORMAT_R32U
								}
						]);
				} else if (!needsNodeIndex && hasNodeIndex) {
						this._format.removeExtraStreams([
								'pcNodeIndex'
						]);
				}
		}
		set lodBehindPenalty(value) {
				if (this._lodBehindPenalty !== value) {
						this._lodBehindPenalty = value;
						this.dirty = true;
				}
		}
		get lodBehindPenalty() {
				return this._lodBehindPenalty;
		}
		set lodRangeMin(value) {
				if (this._lodRangeMin !== value) {
						this._lodRangeMin = value;
						this.dirty = true;
				}
		}
		get lodRangeMin() {
				return this._lodRangeMin;
		}
		set lodRangeMax(value) {
				if (this._lodRangeMax !== value) {
						this._lodRangeMax = value;
						this.dirty = true;
				}
		}
		get lodRangeMax() {
				return this._lodRangeMax;
		}
		set lodUnderfillLimit(value) {
				if (this._lodUnderfillLimit !== value) {
						this._lodUnderfillLimit = value;
						this.dirty = true;
				}
		}
		get lodUnderfillLimit() {
				return this._lodUnderfillLimit;
		}
		set splatBudget(value) {
				if (this._splatBudget !== value) {
						this._splatBudget = value;
						this.dirty = true;
				}
		}
		get splatBudget() {
				return this._splatBudget;
		}
		set colorRamp(value) {
				if (this._colorRamp !== value) {
						this._colorRamp = value;
						this.dirty = true;
				}
		}
		get colorRamp() {
				return this._colorRamp;
		}
		set dataFormat(value) {
				if (this._dataFormat !== value) {
						this._dataFormat = value;
						const extraStreams = this._format.extraStreams.map((s)=>({
										name: s.name,
										format: s.format,
										storage: s.storage
								}));
						this._format = this._createFormat(value);
						if (extraStreams.length > 0) {
								this._format.addExtraStreams(extraStreams);
						}
						this.dirty = true;
				}
		}
		get dataFormat() {
				return this._dataFormat;
		}
		get material() {
				return this._material;
		}
		get format() {
				return this._format;
		}
		frameEnd() {
				this._material.dirty = false;
				this.dirty = false;
		}
		constructor(device){
				this._material = new ShaderMaterial();
				this._dataFormat = GSPLATDATA_COMPACT;
				this.debugAabbs = false;
				this.radialSorting = false;
				this._gpuSorting = false;
				this.debugNodeAabbs = false;
				this.dirty = false;
				this._colorizeLod = false;
				this._enableIds = false;
				this._culling = false;
				this.lodUpdateDistance = 1;
				this.lodUpdateAngle = 0;
				this._lodBehindPenalty = 1;
				this._lodRangeMin = 0;
				this._lodRangeMax = 10;
				this._lodUnderfillLimit = 0;
				this._splatBudget = 0;
				this._colorRamp = null;
				this.colorRampIntensity = 1;
				this.colorizeColorUpdate = false;
				this.colorUpdateDistance = 0.2;
				this.colorUpdateAngle = 2;
				this.colorUpdateDistanceLodScale = 2;
				this.colorUpdateAngleLodScale = 2;
				this.cooldownTicks = 100;
				this._device = device;
				this._format = this._createFormat(GSPLATDATA_COMPACT);
		}
}

export { GSplatParams };
