// playcanvas/build/playcanvas.prf/src/scene/shader-lib/wgsl/chunks/gsplat/compute-gsplat-compact-flag.js
const computeGsplatCompactFlagSource = `
struct FlagUniforms {
	totalSplats: u32,
	textureWidth: u32,
	visWidth: u32,
	totalThreads: u32,
	numWorkgroupsX: u32
};
@group(0) @binding(0) var<uniform> uniforms: FlagUniforms;
@group(0) @binding(1) var pcNodeIndex: texture_2d<u32>;
@group(0) @binding(2) var nodeVisibilityTexture: texture_2d<u32>;
@group(0) @binding(3) var<storage, read_write> flagBuffer: array<u32>;
#ifdef USE_SORTED_ORDER
@group(0) @binding(4) var<storage, read> sortedOrder: array<u32>;
#endif
@compute @workgroup_size({WORKGROUP_SIZE})
fn main(@builtin(global_invocation_id) gid: vec3u) {
	let totalSplats = uniforms.totalSplats;
	let texW = uniforms.textureWidth;
	let visW = uniforms.visWidth;
	let stride = uniforms.totalThreads;
	let threadId = gid.x + gid.y * uniforms.numWorkgroupsX * {WORKGROUP_SIZE}u;
	for (var idx = threadId; idx < totalSplats; idx += stride) {
		#ifdef USE_SORTED_ORDER
			let splatId = sortedOrder[idx];
		#else
			let splatId = idx;
		#endif
		let uv = vec2i(i32(splatId % texW), i32(splatId / texW));
		let nodeIdx = textureLoad(pcNodeIndex, uv, 0).r;
		let texelIdx = nodeIdx >> 5u;
		let bitIdx = nodeIdx & 31u;
		let visCoord = vec2i(i32(texelIdx % visW), i32(texelIdx / visW));
		let visBits = textureLoad(nodeVisibilityTexture, visCoord, 0).r;
		flagBuffer[idx] = select(0u, 1u, (visBits & (1u << bitIdx)) != 0u);
	}
	if (threadId == 0u) {
		flagBuffer[totalSplats] = 0u;
	}
}
`;

export { computeGsplatCompactFlagSource };
