// playcanvas/build/playcanvas.prf/src/scene/shader-lib/wgsl/chunks/gsplat/compute-gsplat-interval-cull.js
const computeGsplatIntervalCullSource = `
struct Interval {
	workBufferBase: u32,
	splatCount: u32,
	boundsIndex: u32,
	pad: u32
};
struct CullUniforms {
	numIntervals: u32,
	visWidth: u32
};
@group(0) @binding(0) var<uniform> uniforms: CullUniforms;
@group(0) @binding(1) var<storage, read> intervals: array<Interval>;
@group(0) @binding(2) var<storage, read_write> countBuffer: array<u32>;
#ifdef CULLING_ENABLED
@group(0) @binding(3) var nodeVisibilityTexture: texture_2d<u32>;
#endif
@compute @workgroup_size({WORKGROUP_SIZE})
fn main(@builtin(global_invocation_id) gid: vec3u) {
	let idx = gid.x;
	if (idx < uniforms.numIntervals) {
		let interval = intervals[idx];
		#ifdef CULLING_ENABLED
			let boundsIdx = interval.boundsIndex;
			let texelIdx = boundsIdx >> 5u;
			let bitIdx = boundsIdx & 31u;
			let visW = uniforms.visWidth;
			let visCoord = vec2i(i32(texelIdx % visW), i32(texelIdx / visW));
			let visBits = textureLoad(nodeVisibilityTexture, visCoord, 0).r;
			let visible = (visBits & (1u << bitIdx)) != 0u;
			countBuffer[idx] = select(0u, interval.splatCount, visible);
		#else
			countBuffer[idx] = interval.splatCount;
		#endif
	}
	if (idx == 0u) {
		countBuffer[uniforms.numIntervals] = 0u;
	}
}
`;

export { computeGsplatIntervalCullSource };
